src/components/SourcesModal.js

```diff
@@ -0,0 +1,208 @@
+import { bindActionCreators } from "redux";
+import { filter } from "fuzzaldrin-plus";
+
+import actions from "../actions";
+import {
+  getSources,
+  getActiveSearch,
+  getSourceSearchQuery
+} from "../selectors";
+import { endTruncateStr } from "../utils/utils";
+import { scrollList } from "../utils/result-list";
+import { isPretty, getSourcePath, isThirdParty } from "../utils/source";
+import Modal from "./shared/Modal";
+import SearchInput from "./shared/SearchInput";
+import ResultList from "./shared/ResultList";
+
+import type { SourcesMap } from "../reducers/sources";
+
+function formatResults(sources: SourcesMap) {
+  return sources
+    .valueSeq()
+    .toJS()
+    .filter(source => !isPretty(source) && !isThirdParty(source))
+    .map(source => ({
+      value: getSourcePath(source),
+      title: getSourcePath(source)
+        .split("/")
+        .pop(),
+      subtitle: endTruncateStr(getSourcePath(source), 100),
+      id: source.id
+    }))
+    .filter(formattedSource => formattedSource.value != "");
+}
+
+type Props = {
+  enabled: boolean,
+  sources: Array<Object>,
+  query: string,
+  selectSource: (id: string) => void,
+  setSourceSearchQuery: (query: string) => void,
+  closeActiveSearch: () => void
+};
+
+type State = {
+  results: ?Array<Object>,
+  selectedIndex: number
+};
+
+class SourcesModal extends Component {
+  props: Props;
+  state: State;
+  constructor(props) {
+    super(props);
+    this.state = { results: null, selectedIndex: 0 };
+  }
+
+  componentDidMount() {
+    this.searchSources(this.props.query);
+  }
+
+  componentDidUpdate(prevProps: any) {
+    if (this.refs.resultList && this.refs.resultList.refs) {
+      scrollList(this.refs.resultList.refs, this.state.selectedIndex);
+    }
+
+    if (!prevProps.enabled && this.props.enabled) {
+      this.searchSources(this.props.query);
+    }
+  }
+
+  closeModal = () => {
+    this.props.closeActiveSearch();
+  };
+
+  searchSources = (query: string) => {
+    if (query == "") {
+      this.setState({ results: this.props.sources });
+      return;
+    }
+    this.setState({
+      results: filter(this.props.sources, query, { key: "value" })
+    });
+  };
+
+  resultCount() {
+    return this.state.results ? this.state.results.length : 0;
+  }
+
+  selectResultItem = (e: SyntheticEvent, result) => {
+    const { selectSource } = this.props;
+
+    if (!result) {
+      return;
+    }
+
+    selectSource(result.id);
+
+    this.closeModal();
+  };
+
+  traverseResults = (direction: number) => {
+    const { selectedIndex } = this.state;
+    const resultCount = this.resultCount();
+    const index = selectedIndex + direction;
+    const nextIndex = (index + resultCount) % resultCount;
+
+    this.setState({ selectedIndex: nextIndex });
+  };
+
+  onChange = (e: SyntheticInputEvent) => {
+    this.props.setSourceSearchQuery(e.target.value);
+    this.searchSources(e.target.value);
+  };
+
+  onKeyDown = (e: SyntheticKeyboardEvent) => {
```

Comment:
By using `onKeyDown` instead of `onKeyUp` we can let the user hold down a key to continue scrolling through a list.